### PR TITLE
Re-design Indices lowering

### PR DESCRIPTION
**Pull request by yzh119**
Previously the indices lowering stage is not flexible not not supporting indices that are not sparse itervars and one sparse itervar can only be attached to a single sparse buffer, also multiple sparse-blocks is not supported, which is not reasonable in the practical case.

These limitations hinders non-zero elements inference and more examples such as sparse softmax. This PR fix the issue by re-designing the indices lowering algorithm.

More specifically, this PR completely removes analyzer (that collects dependency information), but use a sparse block context stack to store the sparse block local information used for rewriting, also, a sparse iter var is no longer attached to a sparse buffer.

Support indices more complicated than sparse itervars is supported by a visitor that recursively transforms var nodes.

UPD: unittests passed.

TODOs:
- [ ] Docstring not ready.
- [ ] Decompress offset for sparse axis from coordinates not finished yet (but it won't affect unittest).
- [ ] More tests involves dense iterator viewed in sparse axis.

These TODOs will be implemented in following PRs.

### CODE REVIEW

**Comment by yzh119** [comment]
@MasterJH5574 Let's sync after Dec 15th.
@junrushao1994 any code-wise suggestions?

File: src/tir/transforms/lower_sparse_tir.cc [hunk 1]
```diff
@@ -83,178 +83,217 @@ Map<Var, Buffer> UpdateBufferMap(PrimFunc f) {
...
+    // initialize sparse iter var dependency map.
+    std::unordered_map<String, SpIterVar> axis_name_sp_iter_map_;
+    for (const SpIterVar& sp_iter_var : sp_iter_vars) {
+      axis_name_sp_iter_map_[sp_iter_var->axis->name] = sp_iter_var;
+    }
+
+    for (const SpIterVar& sp_iter_var : sp_iter_vars) {
+      String axis_name = sp_iter_var->axis->name;
+      const SpIterVarNode* node = sp_iter_var.get();
+      if (axis_name.size() > 6 && std::string(axis_name).compare(axis_name.size() - 6, 6, "_dense") == 0) {
```
Comment (junrushao1994) [inline src/tir/transforms/lower_sparse_tir.cc:101]: Use `EndsWith` defined in src/support/utils.h

File: src/tir/transforms/lower_sparse_tir.cc [hunk 2]
```diff
@@ -83,178 +83,217 @@ Map<Var, Buffer> UpdateBufferMap(PrimFunc f) {
...
+          CHECK(it != axis_name_sp_iter_map_.end())
+              << "Cannot find sparse iter vars corresponding to parent axis " << parent_axis_name
+              << " in current sparse block " << blk_name_;
+          parent_sp_iter_var_[node] = it->second;
+        }
+      }
+    }
+  }
+
+  void add_sp_iter(const VarNode* node, const SpIterVar& sp_iter_var) {
```
Comment (junrushao1994) [inline src/tir/transforms/lower_sparse_tir.cc:119]: CamelCase for methods

```suggestion
  void AddSpIter(const VarNode* node, const SpIterVar& sp_iter_var) {
```

**Comment by yzh119** [comment]
Thanks, code revised according to suggestions.

**Pull request merged**